waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -0,0 +1,76 @@
+## Nim-LibP2P
+## Copyright (c) 2020 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import std/[oids, options]
+import chronos
+import chronicles
+import bearssl
+import nimcrypto/[utils, sha2, hmac]
+
+import libp2p/stream/[connection]
+import libp2p/protobuf/minprotobuf
+import libp2p/utility
+import libp2p/errors
+import libp2p/crypto/[crypto, chacha20poly1305]
+
+when defined(libp2p_dump):
+  import libp2p/debugutils
+
+logScope:
+  topics = "nim-waku noise"
+
+const
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey = default(ChaChaPolyKey)
+
+type
+  ChaChaPolyCiphertext* = object
+    data: seq[byte]
+    tag: ChaChaPolyTag
+
+  ChaChaPolyCipherState* = object
+    k*: ChaChaPolyKey
+    nonce*: ChaChaPolyNonce
+    ad*: seq[byte]
+
+  NoiseError* = object of LPError
+  NoiseDecryptTagError* = object of NoiseError
+
+# ChaChaPoly encryption
+proc encrypt*(
+    state: ChaChaPolyCipherState,
+    plaintext: openArray[byte]): ChaChaPolyCiphertext
+    {.noinit, raises: [Defect].} =
+  #TODO: add padding
```

Comment:
Can we raise a Github issue for this maybe?

Reply:
OK not to address straight away, as long as we capture it somewhere as it may impact security

Reply:
Sure, in fact padding is a requirement we have for encrypted after-handshake transport messages. Here we're still adding vanilla ChaChaPoly encryption, so padding is not necessarily a requirement, but it would be nice to have it as an option later on.